kernel/src/checkpoint/mod.rs

```diff
@@ -1,8 +1,481 @@
+            counts_tx: self.counts_tx.clone(),
+            actions_count: 0,
+            add_actions_count: 0,
+            fully_consumed: false,
+        };
+
+        Ok(CheckpointData {
+            path: checkpoint_path.location,
+            data: wrapped_iterator,
+        })
+    }
+
+    /// TODO(#850): Implement the finalize method
+    ///
+    /// Finalizes checkpoint creation after verifying all data is persisted.
+    ///
+    /// This method **must** be called only after:
+    /// 1. The checkpoint data iterator has been fully consumed
+    /// 2. All data has been successfully written to object storage
+    ///
+    /// # Parameters
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `metadata`: A single-row, single-column [`EngineData`] batch containing:
+    ///   - `sizeInBytes` (i64): The size of the written checkpoint file
+    ///
+    /// # Returns: [`variant@Ok`] if the checkpoint was successfully finalized
+    #[allow(unused)]
+    fn finalize(self, _engine: &dyn Engine, _metadata: &dyn EngineData) -> DeltaResult<()> {
+        // The method validates iterator consumption, but can not guaruntee data persistence.
+        match self.counts_rx.try_recv() {
+            Ok(counts) => {
+                // Write the _last_checkpoint file with the action counts
+                todo!("Implement the finalize method which will write the _last_checkpoint file")
+            }
+            Err(_) => {
+                // The iterator wasn't fully consumed, which means not all data was written
+                Err(Error::checkpoint_write(
+                    "Checkpoint data iterator was not fully consumed before finalization",
+                ))
+            }
+        }
+    }
+
+    /// Creates the checkpoint metadata action for V2 checkpoints.
+    ///
+    /// This function generates the [`CheckpointMetadata`] action that must be included in the
+    /// V2 spec checkpoint file. This action contains metadata about the checkpoint, particularly
+    /// its version.
+    ///
+    /// # Implementation Details
+    ///
+    /// The function creates a single-row [`EngineData`] batch containing only the
+    /// version field of the [`CheckpointMetadata`] action. Future implementations will
+    /// include the additional metadata field `tags` when map support is added.
+    ///
+    /// # Returns:
+    /// A [`CheckpointBatch`] batch including the single-row [`EngineData`] batch along with
+    /// an accompanying selection vector with a single `true` value, indicating the action in
+    /// batch should be included in the checkpoint.
+    fn create_checkpoint_metadata_batch(
+        &self,
+        version: i64,
+        engine: &dyn Engine,
+    ) -> DeltaResult<CheckpointBatch> {
+        let values: &[Scalar] = &[version.into()];
+
+        let checkpoint_metadata_batch = engine
+            .evaluation_handler()
+            .create_one(CHECKPOINT_METADATA_ACTION_SCHEMA.clone(), values)?;
```

Comment:
This might work, but I suspect fmt will decide it's "too complex" and inject line breaks?
```suggestion
            .create_one(CHECKPOINT_METADATA_ACTION_SCHEMA.clone(), &[Scalar::from(version)])?;
```

Reply:
It did, but it looks cleaner than separating out values variable imo